File: common/src/main/java/lighthouse/files/DirectoryWatcher.java
package lighthouse.files;

import com.google.common.collect.ImmutableSet;
import com.sun.nio.file.SensitivityWatchEventModifier;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Nullable;
import java.io.IOException;
import java.nio.file.*;
import java.util.concurrent.Executor;
import java.util.function.BiConsumer;

import static java.nio.file.StandardWatchEventKinds.*;

// TODO: This class should buffer up and deduplicate changes to avoid a stream of MODIFY MODIFY MODIFY events when a file is being copied in to a watched area.
// The rest of the code does handle this, but it requires special cases in a bunch of places. In general this class
// could be a whole lot easier to use.

/**
 * The Java directory watching API is very low level, almost a direct translation of the underlying OS API's, so we
 * wrap it here to make it more digestable.
 */
public class DirectoryWatcher {
    private static final Logger log = LoggerFactory.getLogger(DirectoryWatcher.class);
    private final Thread thread;

    public DirectoryWatcher(ImmutableSet<Path> directories, BiConsumer<Path, WatchEvent.Kind<Path>> onChanged, @Nullable Executor executor) {
        if (directories.isEmpty()) {
            thread = null;
            return;
        }
        log.info("Starting directory watch service for {}", directories);
        thread = new Thread(() -> {
            try {
                WatchService watcher = FileSystems.getDefault().newWatchService();
                for (Path directory : directories) {
                    directory.register(watcher, new WatchEvent.Kind[]{ENTRY_DELETE, ENTRY_CREATE, ENTRY_MODIFY},
                            SensitivityWatchEventModifier.HIGH);
                }
                while (!Thread.currentThread().isInterrupted()) {
                    WatchKey key = watcher.take();
                    for (WatchEvent<?> event : key.pollEvents()) {
                        WatchEvent.Kind<?> kind = event.kind();
                        if (kind == OVERFLOW) {
                            continue;
                        }
                        WatchEvent<Path> ev = (WatchEvent<Path>) event;
                        Path filename = ((Path)key.watchable()).resolve(ev.context());
                        if (executor != null)
                            executor.execute(() -> onChanged.accept(filename, ev.kind()));
                        else
                            onChanged.accept(filename, ev.kind());
                    }
                    if (!key.reset())
                        break;
                }
            } catch (IOException e) {
                e.printStackTrace();
            } catch (InterruptedException e) {
                // Shutting down ...
            }
        }, "Directory watcher");
        thread.setDaemon(true);
        thread.start();
    }

    public void stop() {
        if (thread != null) thread.interrupt();
    }
}

File: common/src/main/java/lighthouse/files/DirectoryWatcher.kt
package lighthouse.files

import com.sun.nio.file.SensitivityWatchEventModifier
import lighthouse.threading.AffinityExecutor
import lighthouse.utils.ThreadBox
import org.slf4j.LoggerFactory
import java.io.IOException
import java.nio.file.FileSystems
import java.nio.file.Path
import java.nio.file.StandardWatchEventKinds.ENTRY_CREATE
import java.nio.file.StandardWatchEventKinds.ENTRY_DELETE
import java.nio.file.StandardWatchEventKinds.ENTRY_MODIFY
import java.nio.file.StandardWatchEventKinds.OVERFLOW
import java.nio.file.WatchEvent
import java.time.Duration
import java.util.concurrent.ScheduledFuture
import kotlin.concurrent.thread

/**
 * The Java directory watching API is very low level, almost a direct translation of the underlying OS API's, so we
 * wrap it here to make it more digestable. One of the things this does is buffer up rapid sequences of notifications
 * that can be caused by file copy tools like scp.
 */
public object DirectoryWatcher {
    private val log = LoggerFactory.getLogger(DirectoryWatcher::class.java)

    @Suppress("UNCHECKED_CAST") @JvmStatic
    public fun watch(directory: Path, executor: AffinityExecutor.ServiceAffinityExecutor, onChanged: (Path, WatchEvent.Kind<Path>) -> Unit): Thread {
        return thread(start = true, daemon = true, name = "Directory watcher for $directory") {
            log.info("Starting directory watch service for $directory")

            // Apply a short delay to collapse rapid sequences of notifications together.
            class Pending(val kind: WatchEvent.Kind<Path>, val future: ScheduledFuture<*>)

            val dupeMap = ThreadBox(hashMapOf<Path, Pending>())

            try {
                val watcher = FileSystems.getDefault().newWatchService()
                directory.register(watcher, arrayOf(ENTRY_DELETE, ENTRY_CREATE, ENTRY_MODIFY), SensitivityWatchEventModifier.HIGH)
                while (!Thread.currentThread().isInterrupted) {
                    val key = watcher.take()
                    for (event in key.pollEvents()) {
                        val kind = event.kind()
                        if (kind === OVERFLOW) {
                            continue
                        }
                        val ev = event as WatchEvent<Path>
                        val filename = (key.watchable() as Path).resolve(ev.context())

                        val handler = {
                            dupeMap.useWith { remove(filename) }
                            onChanged(filename, ev.kind())
                        }

                        dupeMap.use { map ->
                            val pending: Pending? = map[filename]

                            fun addToMap() {
                                map[filename] = Pending(ev.kind(), executor.executeIn(Duration.ofSeconds(1), handler))
                            }

                            if (pending == null) {
                                addToMap()
                            } else {
                                if (ev.kind() == ENTRY_MODIFY && pending.kind == ENTRY_CREATE) {
                                    // We do nothing here, as the onChanged event will prefer to see a CREATE and not
                                    // the subsequent modifies.
                                } else if (ev.kind() == ENTRY_DELETE && pending.kind == ENTRY_CREATE) {
                                    // A file created and deleted so fast can just be ignored.
                                    pending.future.cancel(false)
                                } else {
                                    // Otherwise let it override the previous pending event.
                                    pending.future.cancel(false)
                                    addToMap()
                                }
                            }
                        }
                    }
                    if (!key.reset())
                        break
                }
            } catch (e: IOException) {
                e.printStackTrace()
            } catch (e: InterruptedException) {
                // Shutting down ...
            }
        }
    }
}
